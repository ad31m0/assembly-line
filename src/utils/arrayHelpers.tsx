/* istanbul ignore file */
/**
 *  Wrapper function for array push (append)
 * 
 * @param stage: string []
 * @param item: string
 * @returns stage: string[]
 */
export const push = (stage: string[], item: string) => {
    stage.push(item);
    return stage;
}
/**
 *  Wrapper function for array unshift (prepend)
 * 
 * @param stage: string []
 * @param item: string
 * @returns stage: string[]
 */
export const unshift = (stage: string[], item: string) => {
    stage.unshift(item);
    return stage;
}
/**
 *  Wrapper function for array remove with given item index
 * 
 * @param stage: string []
 * @param itemIdx: string
 * @returns stage: string[]
 */
export const remove = (stage: string[], itemIdx: number) => {
    stage.splice(itemIdx, 1);
    return stage;
}