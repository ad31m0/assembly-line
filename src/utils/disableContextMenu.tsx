import { MouseEvent } from "react";

export const disableContextMenu = (event: MouseEvent) => {
  if (event.preventDefault !== undefined)
    event.preventDefault();
  if (event.stopPropagation !== undefined)
    event.stopPropagation();
  return false;
};