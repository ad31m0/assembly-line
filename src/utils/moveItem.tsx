
export type moveItemType = (itemIdx: number, stageIdx: number, delta: number, itemName?: string) => void;
/* istanbul ignore next */
export const push = (stage: string[], item: string) => {
    stage.push(item);
    return stage;
}
/* istanbul ignore next */
export const unshift = (stage: string[], item: string) => {
    stage.unshift(item);
    return stage;
}

export const remove = (stage: string[], itemIdx: number) => {
    stage.splice(itemIdx, 1);
    return stage;
}

export const moveItemCtor = ({ stagesItems, setStagesItems }: { stagesItems: Array<string[]>, setStagesItems: (items: Array<string[]>) => void }) => {
    const moveItem: moveItemType = (itemIdx, stageIdx, delta, itemName) => {
        
        if (stageIdx > -1 && stageIdx < stagesItems.length) {
            itemName = stagesItems[stageIdx][itemIdx];
            stagesItems[stageIdx] = remove(stagesItems[stageIdx], itemIdx);
        }
        if ((stageIdx === 0 && delta < 0) || (stageIdx === stagesItems.length - 1 && delta > 0)) {
            // remove from pipeline don't add to another stage
        } else {
            const stage = stagesItems[stageIdx + delta];
            if (delta > 0) {
                stagesItems[stageIdx + delta] = push(stage, itemName as string);
            } else {
                stagesItems[stageIdx + delta] = unshift(stage, itemName as string);
            }
        }
        setStagesItems([...stagesItems]);
    }
    return moveItem;
}