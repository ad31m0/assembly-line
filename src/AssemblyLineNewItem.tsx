import { useState } from 'react';
import { moveItemType } from "./utils/moveItem";

export const AssemblyLineNewItem = ({ moveItem }: { moveItem: moveItemType }) => {
  const [newItemName, setNewItemName] = useState("");
  return (<div>
    <span>Add an item: </span>
    <input
      maxLength={160}
      value={newItemName}
      onChange={(event) => {
        setNewItemName(event.target.value);
      }}
      onKeyPress={(event) => {
        if (event.code === "Enter" && newItemName !== "") {
          moveItem(-1, -1, +1, newItemName);
          setNewItemName("");
        }
      }}
    />
  </div>)
}  