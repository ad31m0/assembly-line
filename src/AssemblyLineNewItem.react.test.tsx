import React from 'react';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { configure, shallow } from 'enzyme';
import { AssemblyLineNewItem } from './AssemblyLineNewItem';
import { moveItemType } from './utils/moveItem';

configure({ adapter: new Adapter() });

describe('<AssemblyLine /> with no props', () => {
  const moveItemMock = jest.fn() as jest.MockedFunction<moveItemType>;
  moveItemMock.mockImplementation(() => { });
  const container = shallow(<AssemblyLineNewItem moveItem={moveItemMock} />);
  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot();
  });

  it('should have a add new item input', () => {
    expect(container.find('input').length).toEqual(1);
  });

  it('add new item should only save on enter', () => {
    container.find('input').simulate('change', { target: { value: 'item1' } });
    expect(container.find('input').prop('value')).toEqual('item1');

    container.find('input').simulate('keypress', { code: "anyother" });
    expect(moveItemMock).not.toHaveBeenCalled();

    container.find('input').simulate('keypress', { code: "Enter" });
    expect(moveItemMock).toBeCalledWith(-1, -1, +1, "item1");
    expect(container.find('input').prop('value')).toEqual('');

  });

  it('should not save empty', () => {
    container.find('input').simulate('change', { target: { value: '' } });
    expect(container.find('input').prop('value')).toEqual('');

    container.find('input').simulate('keypress', { code: "Enter" });
    expect(moveItemMock).not.toHaveBeenCalled();
  });

});
