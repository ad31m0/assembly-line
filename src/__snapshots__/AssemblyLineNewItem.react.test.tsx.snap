// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssemblyLine /> with no props should match the snapshot 1`] = `"<div><span>Add an item: </span><input maxLength=\\"160\\" value=\\"\\"/></div>"`;
